Debugger (Debug Adapter Protocol)

bluesign opened the pull request:
## Description

I couldn't manage to push over existing PR, so made a new one.  /cc @turbolent 

Making this draft PR to track also issues. 

<img width="852" alt="image" src="https://user-images.githubusercontent.com/1119338/214288938-2348b236-50f4-4ecd-b323-2ab033e553b6.png">

Current Problems :
- Pre/Post conditions has wrong stack ( not sure how to solve this one for sure ) 
- Figure out how to evaluate statement ( not variables ) 

Missing: 
- Transaction Execution ( can be solved by attaching & executing transaction with plugin or cli ) 
- Arguments ( I think this is also not in the context of debugger ) 

Also:
- there is also an option to utilize --skip-tx-validation for emulator, and put authorizer addresses in some pragma , we can use proposer and payer as service account
- Arguments can be pragma too 

Wishlist:
- Debug mainnet script / transaction  ( hard one but I think can be managed, main problem is the lifetime of registers vs debugging session length ( maybe archival node can help there )  ) 

Done:
- [DONE] Stringer broken for script self
- [DONE] Stringer broken for destroyed resources 
- [DONE] Variables include global functions when in interface post condition 
- [DONE] Better Variables ( expand to see member fields ) 
- [DONE] Step Over & In & Out
- [DONE] Attaching debugger ( scripts )
- [DONE] expand arrays, dictionaries
- [DONE] add debugger to vscode cadence extension 
- [DONE] Storage Browser
- [DONE] Attaching debugger (transactions)
- [DONE] need to hook on `log` (waiting logger fix PR )
- [DONE] handle StorageReference in variables (borrowed resources) 
______

For contributor use:

- [X] Targeted PR against `master` branch
- [ ] Linked to GitHub issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation
- [X] Re-reviewed `Files changed` in the GitHub PR explorer
- [ ] Added appropriate labels

bluesign commented:
> Attaching debugger (transactions): bug in flow-go ( using env config instead of runtime or something like that, doesn't happen with script, as emulator/fvm doesn't execute script before us )

@pattyshack any idea on this one ?

bluesign commented:
@turbolent 

> Pre/Post conditions has wrong stack ( not sure how to solve this one for sure )

These are coming with no stack except their self. 

> Figure out how to evaluate statement ( not variables )

Is it possible to evaluate a string ( as code ) via interpreter. 

> need to hook on log

logs are coming after execution; I can bug also fvm for this ( it would be nice to have in emulator too )

turbolent commented:
Thank you for taking continuing this effort @bluesign! 

The original PR, #164, was already quite large and it would be better/ easier to get all the functionality landed in pieces.

Do you think there is a commit at which point the basic functionality from the original PR was restored? Maybe [f8631b2](https://github.com/onflow/flow-emulator/pull/294/commits/f8631b2be68e2854bcc9fa5234cf3bf3da4b482f)? We could maybe get those commits merged back into the original PR and get that merged, then open another for subsequent features / improvements you added (👏).

turbolent commented on blockchain.go:
```diff
@@ -931,8 +938,10 @@ func (b *Blockchain) ExecuteNextTransaction() (*types.TransactionResult, error)
+		fvm.WithReusableCadenceRuntimePool(
+			reusableRuntime.NewCustomReusableCadenceRuntimePool(1, func(runtime.Config) runtime.Runtime {
+				return b.cadenceRuntime
+			})))
```
The code for building a new FVM context from a parent, given a header and the FVM options, is duplicated in `executeBlock` and `ExecuteScriptAtBlock`. Maybe refactor this into a function which is called in both places, so the FVM options stay consistent.

bluesign replied:
this should be changed later on commits. `fvm.WithReusableCadenceRuntimePool` not needed anymore, I moved the debugger to initialization to global ( I had to actually, because `NewCustomReusableCadenceRuntimePool` loses config somehow, only way to use is `NewReusableCadenceRuntimePool` to keep debugger ) Long story short, I think it is not duplicated anymore there.

turbolent commented on blockchain.go:
```diff
@@ -1207,6 +1225,11 @@ func (b *Blockchain) ExecuteScriptAtBlock(script []byte, arguments [][]byte, blo
 	}, nil
 }
 
+func ComputeScriptID(script []byte) sdk.Identifier {
+	hasher := hash.NewSHA3_256()
+	return sdk.HashToID(hasher.ComputeHash(script))
+}
```
Not sure, but maybe the SDK or flow-go/FVM already expose this? It would be good to have it there, so that this code stays in sync / this duplication does not get out sync.

turbolent commented on blocks.go:
```diff
@@ -68,6 +66,10 @@ func (b *blocks) FinalizedHeader() (*flowgo.Header, error) {
-		ByHeightFrom(height, header)
+	block, err := b.blockchain.storage.BlockByHeight(context.Background(), height)
+	if err != nil {
+		return nil, err
+	}
+
+	return block.Header, nil
```
Was there something wrong with the old approach?

bluesign replied:
this is related to #280

turbolent reviewed (commented):
Reviewed all commits up to f8631b2, looks good!

bluesign commented:
> Do you think there is a commit at which point the basic functionality from the original PR was restored? Maybe [f8631b2](https://github.com/onflow/flow-emulator/pull/294/commits/f8631b2be68e2854bcc9fa5234cf3bf3da4b482f)? We could maybe get those commits merged back into the original PR and get that merged, then open another for subsequent features / improvements you added (👏).

@turbolent Yeah this is totally possible, my only concern was, sometimes when adding a feature you have to go back and change everything from ground up. ( I had that with attaching debugger for example ) Instead of attaching, we had to attach one and then reuse for a clean implementation and attach support. 

I think we have only 2 questions left to answer, then I can convert to WIP PR to mini PRs. 

- Pre/Post conditions has wrong stack ( they don't have stack on interface post/pre conditions ) 
- Figure out how to evaluate statement ( like self.x, or any statement, for watches window )
- Maybe also adding globals to variable ( this may need cadence change )

If those 2 doesn't change something in the architecture, I can start making mini feature PRs.

turbolent commented:
@bluesign The features you listed are great ideas for additions, but they are maybe better added in follow-up PRs. What is left to be done for the current functionality?

bluesign commented:
> @bluesign The features you listed are great ideas for additions, but they are maybe better added in follow-up PRs. What is left to be done for the current functionality?

@turbolent  In my todo list, for missing stuff and bugs: 

- `resource after move` causing a crash 

 I can simply ignore this one, by recovering panic I guess, I just need to check. 

- transaction pre/post nested transactions 

when executing transactions fvm also runs some stuff (fees, storage check etc); we debug them too :) need to find a way to skip them. 

- transaction and scripts with arguments 

I have implemented attach, and it is working very well when I launch them from CLI, etc. So maybe launching part, we can move to vscode-extension.  ( emulator also has a skipping tx validation option, so without keys we can manage ) 

- Storage and variables working without problem, there are small issues like references not showing & in front as type, etc. 

- I cleaned up the code today, but maybe I need to make smaller PRs, as we discussed.

bluesign commented:
> Integrating with the CLI (FlowKit) to leverage the flow.json configuration to map on-chain addresses to local files

Also, I have a plan for this one, but I think first, I need to discuss with @sideninja about the road plan.

bluesign commented:
> > @bluesign The features you listed are great ideas for additions, but they are maybe better added in follow-up PRs. What is left to be done for the current functionality?
> 
> @turbolent In my todo list, for missing stuff and bugs:
> 
> * `resource after move` causing a crash
> 
> I can simply ignore this one, by recovering panic I guess, I just need to check.
> 
> * transaction pre/post nested transactions
> 
> when executing transactions fvm also runs some stuff (fees, storage check etc); we debug them too :) need to find a way to skip them.
> 
> * transaction and scripts with arguments
> 
> I have implemented attach, and it is working very well when I launch them from CLI, etc. So maybe launching part, we can move to vscode-extension. ( emulator also has a skipping tx validation option, so without keys we can manage )
> 
> * Storage and variables working without problem, there are small issues like references not showing & in front as type, etc.
> * I cleaned up the code today, but maybe I need to make smaller PRs, as we discussed.

I would say we ignore crash for first one and sync with vscode-extension team for the rest. Rest can be done in upcoming PRs. 

I can split this PR if needed ( I made it already half of the previous when cleaning code )

turbolent commented:
@bluesign That is a good list for additional features, but I meant more like, is the PR in its current state already usable or is there still some work left to be done?

bluesign commented:
> @bluesign That is a good list for additional features, but I meant more like, is the PR in its current state already usable or is there still some work left to be done?

Yeah it is currently already usable and pretty useful

turbolent commented:
I synced with master and adjusted the logging to it, then tried debugging a script from VS Code and it worked like a charm – awesome! Let's get this in and then we can open follow-up PRs for the remaining features

turbolent commented on blockchain.go:
```diff
@@ -762,6 +787,45 @@ func (b *Blockchain) GetTransactionResult(ID sdk.Identifier) (*sdk.TransactionRe
+	return &sdkAccount, nil
+}
+
+func (b *Blockchain) GetAccountUnsafe(address sdk.Address) (*sdk.Account, error) {
```
This seems to be a duplicate of `GetAccount` without the locking. Why is it, and not `GetAccount`, called from the debug session?

bluesign replied:
Emulator locking when executing script, and debugger trying to access accounts when stepping over, it was causing deadlock.

turbolent replied:
Sounds good. I'll add a comment/TODO

turbolent commented on blockchain.go:
```diff
@@ -1177,21 +1245,20 @@ func (b *Blockchain) ExecuteScriptAtBlock(script []byte, arguments [][]byte, blo
 
-	hasher := hash.NewSHA3_256()
-	scriptID := sdk.HashToID(hasher.ComputeHash(script))
+	scriptID := sdk.Identifier(flowgo.MakeIDFromFingerPrint(script))
```
👍

turbolent commented on server/debugsession.go:
```diff
@@ -0,0 +1,836 @@
+
+			for name, variable := range functionValues {
+
+				if location.String() == ds.scriptID && name == "self" {
+					continue
+				}
+				if name == "BLS" || name == "RLP" {
+					continue
+				}
```
It would be good to generalize that a bit more, so it stays in sync with Cadence

bluesign replied:
Yeah this one (especially self in script context) I spent too long time digging. 

Maybe we can add few interfaces to Value for making debugging part easier.

turbolent replied:
Sounds good. Adding a TODO

turbolent commented on server/debugsession.go:
```diff
@@ -0,0 +1,836 @@
+		}
+
+	case cadence.Array:
+		i := 0
+		for _, element := range value.Values {
+			variable := ds.convertValueToDapVariable(fmt.Sprintf("[%d]", i), element)
+			if variable != nil {
+				members = append(
+					members,
+					*variable)
+			}
+			i = i + 1
+		}
```
```suggestion
		for i, element := range value.Values {
			variable := ds.convertValueToDapVariable(fmt.Sprintf("[%d]", i), element)
			if variable != nil {
				members = append(
					members,
					*variable
				)
			}
		}
```

turbolent commented on server/debugsession.go:
```diff
@@ -0,0 +1,836 @@
+	return members
+}
+
+func (ds *debugSession) convertMembersToDapVariables(inter *interpreter.Interpreter, someValue any) []dap.Variable {
```
Maybe instead of having this "catch-all function", have a separate function per type and call the appropriate

turbolent replied:
Refactored it in [fabd27d](https://github.com/onflow/flow-emulator/pull/294/commits/fabd27d41700b6ed90d28d0c6c0eb1a1245fcae0)

bluesign commented:
@turbolent thank you very much

turbolent commented on server/backend/emulator.go:
```diff
@@ -39,7 +43,11 @@ type Emulator interface {
 	GetCollection(colID sdk.Identifier) (*sdk.Collection, error)
 	GetTransaction(txID sdk.Identifier) (*sdk.Transaction, error)
 	GetTransactionResult(txID sdk.Identifier) (*sdk.TransactionResult, error)
+	//for debugger
+	GetAccountUnsafe(address sdk.Address) (*sdk.Account, error)
```
```suggestion
	// Deprecated: Needed for the debugger right now, do NOT use for other purposes.
	// TODO: refactor
	GetAccountUnsafe(address sdk.Address) (*sdk.Account, error)
```

turbolent commented on blockchain.go:
```diff
@@ -762,6 +787,45 @@ func (b *Blockchain) GetTransactionResult(ID sdk.Identifier) (*sdk.TransactionRe
+	return &sdkAccount, nil
+}
+
+func (b *Blockchain) GetAccountUnsafe(address sdk.Address) (*sdk.Account, error) {
```
```suggestion
// Deprecated: Needed for the debugger right now, do NOT use for other purposes.
// TODO: refactor
func (b *Blockchain) GetAccountUnsafe(address sdk.Address) (*sdk.Account, error) {
```

turbolent commented on server/debugsession.go:
```diff
@@ -0,0 +1,828 @@
+
+			for name, variable := range functionValues {
+
+				if location.String() == ds.scriptID && name == "self" {
```
```suggestion
				// TODO: generalize exclusion of built-ins / standard library definitions

				if location.String() == ds.scriptID && name == "self" {
```

turbolent reviewed (approved):
Great work! 👏 🙌 

I think we can get this in as-is and then open up follow-up PRs for cleaning up some parts and adding additional functionality.

Thank you very much @bluesign for getting it back to work, and also extending it with super cool features like the storage inspection 👌

DylanTinianov reviewed (approved):
This looks great, very excited to get this feature in!

sideninja commented on blockchain.go:
```diff
@@ -592,6 +603,20 @@ func configureTransactionValidator(conf config, blocks *blocks) *access.Transact
 	)
 }
 
+func (b *Blockchain) newFvmContextFromHeader(header *flowgo.Header) fvm.Context {
```
```suggestion
func (b *Blockchain) newFVMContextFromHeader(header *flowgo.Header) fvm.Context {
```

turbolent replied:
Renamed in 230f39716dbbf4881379170235e520fb9d7a9e0f

sideninja commented on server/debugsession.go:
```diff
@@ -0,0 +1,834 @@
+	}
+}
+
+func (ds *debugSession) convertValueToDapVariable(name string, value cadence.Value) dap.Variable {
```
```suggestion
func (ds *debugSession) convertValueToDAPVariable(name string, value cadence.Value) dap.Variable {
```

turbolent replied:
Renamed this and related functions in 97ec31a7ec3502b4cad10b5c5822a909224f33ad

sideninja commented on server/debugsession.go:
```diff
@@ -0,0 +1,834 @@
+			continue
+		}
+
+		startPos := locationRange.StartPosition()
+		endPos := locationRange.EndPosition(nil)
+
+		locationString := locationPath(location)
+		if location.String() == ds.scriptID {
+			locationString = ds.scriptLocation.String()
+		}
+		stackFrames = append(
+			stackFrames,
+			dap.StackFrame{
+				Source: dap.Source{
+					Path: locationString,
+				},
+				Line:      startPos.Line,
+				Column:    startPos.Column + 1,
+				EndLine:   endPos.Line,
+				EndColumn: endPos.Column + 2,
+			},
+		)
```
I feel this could be extracted to a function as it repeats the above logic, and accept range and location and produce stack frame.

turbolent replied:
Good point! 

Refactored in 3a6ae2e638fc6e10f0caba138c7ad76e6232cd20

sideninja commented on server/debugsession.go:
```diff
@@ -0,0 +1,834 @@
+
+type debugSession struct {
+	logger                *zerolog.Logger
+	backend               *backend.Backend
```
Also then you don't have to type assert to a `emulator.Blockchain`

turbolent replied:
Good idea! Maybe something that can be done in a follow-up PR

sideninja replied:
Maybe just add a todo.

sideninja commented on blockchain.go:
```diff
@@ -1351,3 +1420,11 @@ func (b *Blockchain) testAlternativeHashAlgo(sig flowgo.TransactionSignature, ms
+	b.debugger = debugger
+}
+
+func (b *Blockchain) EndDebugging() {
```
nit: `SetDebugger(nil)`

turbolent replied:
Refactored in 08c266d15d1076e8d8992c7540c9a749225bc5c4

sideninja replied:
I also meant if it makes sense to just reuse the exported SetDebugger function to not actually have additional EndDebugging. Highly subjective so feel free to dismiss or even better educate me 😄

sideninja commented on server/debugsession.go:
```diff
@@ -0,0 +1,834 @@
+func (ds *debugSession) dispatchRequest(request dap.Message) {
+	switch request := request.(type) {
+
+	case *dap.InitializeRequest:
```
nit: I feel this could be extracted into functions and then just have a map of type to function call for different requests on the session struct.

turbolent replied:
Yeah, the function become too big. 

Refactored in fdf0093445cf9b9fa18cb794730a5d108ad18434

sideninja commented on server/debugsession.go:
```diff
@@ -0,0 +1,834 @@
+			}
+
+		case ScopeIdentifierStorage:
+			var index int = 1
```
```suggestion
			index := 1
```

turbolent replied:
Refactored in 87c3211102178900f8925d4877b0a60f9ff986bb

sideninja commented on server/debugsession.go:
```diff
@@ -0,0 +1,834 @@
+	return newDAPResponse(request.Seq, request.Command, true)
+}
+
+func newDAPErrorResponse(requestSeq int, command string, message dap.ErrorMessage) *dap.ErrorResponse {
```
nit:
```suggestion
func newDAPErrorResponse(request *dap.Request, message dap.ErrorMessage) *dap.ErrorResponse {
```

turbolent replied:
Refactored in 87c3211102178900f8925d4877b0a60f9ff986bb
